contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,857 @@
+// Bad
+enum Colors {
+  Red,
+  Green,
+  Blue
+}
+
+// Good
+type Colors = 'red' | 'green' | 'blue';
+```
+You can also define a union type by deriving the type from an array. It's a great choice when seeking an easily iterable and directly usable array in code while keeping one source of truth.
+
+```ts
+// Bad
+const COLORS = ['red', 'green', 'blue'] as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { TupleToUnion } from 'type-fest';
+
+const COLORS = ['red', 'green', 'blue'] as const;
+type Colors = TupleToUnion<typeof COLORS>;
+```
+
+Defining union types using an object offers several benefits. This approach it's a great choice when **enum-like identifiers are needed**, while maintaining one source of truth and improving readability.
+
+```ts
+// Bad
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = 'red' | 'green' | 'blue';
+
+// Good
+import { ValueOf } from 'type-fest';
+
+const COLORS = {
+    Red: 'red',
+    Green: 'green',
+    Blue: 'blue',
+} as const;
+type Colors = ValueOf<typeof COLORS]>
+```
+
+If using enum is necessary, utilize string enum. Do not use numeric and heterogenous enums. This can help prevent potential bugs at runtime.
+
+```ts
+// Bad
+enum Colors {
+  Red,
+  Green,
+  Blue,
+}
+enum Colors {
+  Red = 1,
+  Green = 2,
+  Blue = 3,
+}
+
+// Good
+enum Colors {
+  Red = 'red',
+  Green = 'green',
+  Blue = 'blue',
+}
+```
+
+## Usage of type `any` and `unknown`
+
+The `any` type allows assignment to all types and dereference of any property, which is undesirable and should be avoided. Instead, in most cases use the `unknown` type which expresses a similar concept and is much safer as it requires narrowing the type before using it.
+
+```ts
+// Bad
+const danger: any = 'danger';
+danger.invalidProperty // No TypeScript error, but will be `undefined` at runtime
+danger.invalidProperty.nestedInvalidProperty // No TypeScript error, but will throw TypeError at runtime
+
+// Good
+const safer: unknown = 'safer';
+
+// Using `unknown` requires type narrowing before access
+safer.nonExistentProperty // TypeScript error, forces proper type checking
+
+// Narrowing the type before access
+if (typeof safer === 'string') {
+    safer.toUpperCase(); // Allowed after narrowing the type
+}
+```
+
+Usage of `any` is forbidden unless there is a very good reason for it. The `any` type bypass TypeScript type-checking system, thus making the code less safe and more prone to bugs. Any usage of `any` type must be followed with a comment explaining why you are using it.
+
+```ts
+// Bad
+function processAPIResponse(response: any) {
+    // your code
+}
+
+// Good
+// As we are just going to persist the response in Onyx and
+// it would be very hard to type all possible data structures,
+// we just leave `response` as `any`.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function processAPIResponse (response: any) {
+    // your code
+}
+```
```

Comment:
I feel like this section should go right under the "## Usage of type `any` and `unknown`" heading, because it's the most important thing to know in this section.